rosetta/transactions/intent.go

```diff
@@ -0,0 +1,175 @@
+type Intent struct {
+	From     flow.Address
+	To       flow.Address
+	Amount   cadence.UFix64
+	Payer    flow.Address
+	Proposer flow.Address
+
+	ReferenceBlock            flow.Identifier
+	ProposerKeySequenceNumber uint64
+	GasLimit                  uint64
+}
+
+// CreateTransactionIntent creates a transaction Intent from two operations given as input.
+// Specified operations should be symmetrical, a deposit and a withdrawal from two
+// different accounts. At the moment, the only fields taken into account are the
+// account IDs, amounts and type of operation.
+func (p *Parser) CreateTransactionIntent(operations []object.Operation) (*Intent, error) {
+
+	firstOpNegative := strings.HasPrefix(operations[0].Amount.Value, "-")
+	secondOpNegative := strings.HasPrefix(operations[1].Amount.Value, "-")
+
+	// Operations are invalid if both operations have negative amounts.
+	if firstOpNegative && secondOpNegative {
+
+		return nil, failure.InvalidIntent{
+			Description: failure.NewDescription("invalid operations - two deposits specified"),
+			Sender:      operations[0].AccountID.Address,
+			Receiver:    operations[1].AccountID.Address,
+		}
+	}
+
+	// Operations are also invalid if both operations have positive amounts.
+	if !firstOpNegative && !secondOpNegative {
+
+		return nil, failure.InvalidIntent{
+			Description: failure.NewDescription("invalid operations - two withdrawals specified"),
+			Sender:      operations[0].AccountID.Address,
+			Receiver:    operations[1].AccountID.Address,
+		}
+	}
+
+	// Assume the first operation is the one with the negative amount.
+	send := operations[0]
+	receive := operations[1]
+
+	// If that was not the case, switch the send and receive operations.
+	if !strings.HasPrefix(send.Amount.Value, "-") {
+		receive = operations[0]
+		send = operations[1]
+	}
+
+	// Validate the sender and the receiver account IDs.
+	err := p.validate.Account(send.AccountID)
+	if err != nil {
+		return nil, fmt.Errorf("invalid sender account: %w", err)
+	}
+	err = p.validate.Account(receive.AccountID)
+	if err != nil {
+		return nil, fmt.Errorf("invalid receiver account: %w", err)
+	}
+
+	// Validate the currencies specified for deposit and withdrawal.
+	send.Amount.Currency, err = p.validate.Currency(send.Amount.Currency)
+	if err != nil {
+		return nil, fmt.Errorf("invalid sender currency: %w", err)
+	}
+	receive.Amount.Currency, err = p.validate.Currency(receive.Amount.Currency)
+	if err != nil {
+		return nil, fmt.Errorf("invalid receiver currency: %w", err)
+	}
+
+	// Make sure that both the send and receive operations use the same currency.
+	// This is perhaps unnecessary at the moment since we only have a single currency.
+	if send.Amount.Currency != receive.Amount.Currency {
+		return nil, failure.InvalidIntent{
+			Sender:      send.AccountID.Address,
+			Receiver:    receive.AccountID.Address,
+			Description: failure.NewDescription("send and receive currencies do not match"),
+		}
+	}
```

Comment:
It is unnecessary at the moment. However, we should enforce that both of them are `FLOW`, so that should be the check instead.